course5/homework12/comparers.cc

```diff
@@ -0,0 +1,26 @@
+#include <memory>
+#include "comparers.h"
+
+namespace homework12
+{
+std::unique_ptr<AComparer> CreateComparer(bool numeric, bool keyed, unsigned int column, char separator)
+{
+    if (keyed)
+    {
+        if (numeric)
+            return std::make_unique<ColumnComparer<NumericComparer>>(column, separator);
+        else
+            return std::make_unique<ColumnComparer<BasicComparer>>(column, separator);
+    }
+    else
+    {
+        if (numeric)
+            return std::make_unique<NumericComparer>();
+        else
+            return std::make_unique<BasicComparer>();
+
+    }
+}
```

Comment:
Nice pattern: interface + factory building template-customized types. Very nice, but used in a rather inefficient way.

You build the object once, based on one input (the params), but then you pay the cost of calling virtual function O(N log(N)) times!

If, instead, your interface contained the entire call to ```std::sort```, then you'd call the virtual function only once, and then all the comparisons inside of std::sort would be inlined. Consider this!